synapse/handlers/presence.py

```diff
@@ -1254,9 +1255,12 @@ async def set_state(
         if not ignore_status_msg:
             new_fields["status_msg"] = status_msg
 
-        if presence == PresenceState.ONLINE or (
-            presence == PresenceState.BUSY and self._busy_presence_enabled
-        ):
+        if (
+            prev_state.state == PresenceState.OFFLINE
```

Comment:
I'm failing to follow why this is specifically for offline->online; naively I would have expected if it is going from !online->online.

(The busy presence bit seems like it might have the same bug?)

Reply:
I had a reason, which is entirely escaping me right now(and appears to be missing from my notes). Let me test it out, because if this holds up we can remove the 'hacky' bit in `handle_update()`. I *think* it was related to not using `set_presence` on the `/sync` query, as it defaults to 'online'. However, that behavior has recently changed(at least for Element, other clients may not have). Give me some time to look into it again.

Reply:
OK! The various possible state transitions are rather fuzzy in my head, which is why I'm asking. 😄

Reply:
I went ahead and made those changes. No deal(at least as-is). It will need something else to handle the fact that some clients still set presence via the presence endpoint and not through `/sync`.

https://github.com/matrix-org/synapse/actions/runs/5745219286

![presence_temp](https://github.com/matrix-org/synapse/assets/1582365/b84248b1-fa49-461f-8e96-fc8f20b17ce0)

I left my desktop client open(so it should have been syncing in the background, at least until it went to sleep). But then I checked in real quick with my Android mobile, a few minutes later this nonsense started spamming federation.